xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -465,14 +472,20 @@ public Result selectConfig(PickSubchannelArgs args) {
       }
       final String finalCluster = cluster;
       final long hash = generateHash(selectedRoute.routeAction().hashPolicies(), headers);
+      Route finalSelectedRoute = selectedRoute;
       class ClusterSelectionInterceptor implements ClientInterceptor {
         @Override
         public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
             final MethodDescriptor<ReqT, RespT> method, CallOptions callOptions,
             final Channel next) {
-          final CallOptions callOptionsForCluster =
+          CallOptions callOptionsForCluster =
               callOptions.withOption(CLUSTER_SELECTION_KEY, finalCluster)
                   .withOption(RPC_HASH_KEY, hash);
+          if (isTrustedXdsServer
```

Comment:
This is not safe with federation. The listener and routes could come from different control planes. In fact, different routes could come from different control planes. You need to check for this specific resource.